src/Storages/ExternalStream/Kafka/Kafka.cpp

```diff
@@ -466,10 +453,30 @@ Pipe Kafka::read(
         else
             header = storage_snapshot->getSampleBlockForColumns({ProtonConsts::RESERVED_EVENT_TIME});
 
-        auto offsets = getOffsets(*consumer, query_info.seek_to_info, shards_to_query);
+        auto streaming = query_info.syntax_analyzer_result->streaming;
+
+        auto seek_to_info = query_info.seek_to_info;
+        /// seek_to defaults to 'latest' for streaming. In non-streaming case, 'earliest' is preferred.
+        if (!streaming && seek_to_info->getSeekTo().empty())
+            seek_to_info = std::make_shared<SeekToInfo>("earliest");
+
+        auto offsets = getOffsets(*consumer, seek_to_info, shards_to_query);
         assert(offsets.size() == shards_to_query.size());
 
         for (auto [shard, offset] : std::ranges::views::zip(shards_to_query, offsets))
+        {
+            Int64 high_watermark = NO_WARTERMARK;
+            if (!streaming)
+            {
+                auto marks = topic_ptr->queryWatermarks(shard);
+                LOG_INFO(logger, "watermarks low={} high={}", marks.low, marks.high);
+                high_watermark = marks.high;
+
+                if (offset == nlog::EARLIEST_SN)
+                    offset = marks.low;
```

Comment:
if it is not accurate, we can't do this ?

Reply:
This is safe. For example, even in the case that, when it queries Kafka, it gets the low watermark `10`. And when it actually starts reading messages, the low watermark changes to `15` ( because of retention happens ), Start reading from offset `10` is still valid and the first message it gets will be the message with offset `15`.

Reply:
i think it will report error ? no ? otherwise it seems pretty odd. I like to read 10, but you give me 15

Reply:
and i was thinking do we care about low watermark ?

Reply:
> i think it will report error ? no ?
No, it won't. But what I said was not correct either. If the low offset is `15` and it starts reading from `10`, it won't report error, but it will get stuck. So I reverted that change, and do not set `offset` to `marks.low`.

> and i was thinking do we care about low watermark ? when topic is empty, what is the watermark returned ?
Both the low and high watermarks are the same number. For example, a new topic that has no message return `(0, 0)`.

Reply:
I thought if you requested 10 but it has only 15, broker reported error and librdkafka probably captured the error but we didn't consume it ?

Reply:
No, I confirmed there is no error, but it will just keep wait for that message. At least, this is the behavior the version of librdkafka used in proton.